File: legacy/examples/bsc-testnet-smoke/bsc_exact_client.ts
import {
  X402Client,
  X402FetchClient,
  ExactPermitEvmClientMechanism,
  ExactEvmClientMechanism,
  EvmClientSigner,
  SufficientBalancePolicy,
  type SettleResponse,
  decodePaymentPayload,
} from '../../typescript/packages/x402/dist/index.js';
import { createWalletClient, http, type Hex } from 'viem';
import { privateKeyToAccount } from 'viem/accounts';
import { bscTestnet } from 'viem/chains';

type LocalWallet = {
  getAddress(): Promise<string>;
  signMessage(msg: Uint8Array): Promise<string>;
  signTypedData(data: Record<string, unknown>): Promise<string>;
  signTransaction(payload: Record<string, unknown>): Promise<string>;
};

function createLocalWallet(privateKey: Hex): LocalWallet {
  const account = privateKeyToAccount(privateKey);
  return {
    async getAddress() {
      return account.address;
    },
    async signMessage(msg: Uint8Array) {
      return account.signMessage({ message: { raw: msg } });
    },
    async signTypedData(data: Record<string, unknown>) {
      const { domain, types, primaryType, message } = data as {
        domain: Record<string, unknown>;
        types: Record<string, unknown>;
        primaryType: string;
        message: Record<string, unknown>;
      };
      const normalizedTypes = { ...types } as Record<string, unknown>;
      delete normalizedTypes.EIP712Domain;
      return account.signTypedData({
        domain: domain as any,
        types: normalizedTypes as any,
        primaryType: primaryType as any,
        message: message as any,
      });
    },
    async signTransaction(payload: Record<string, unknown>) {
      const client = createWalletClient({
        account,
        chain: bscTestnet,
        transport: http(process.env.BSC_TESTNET_RPC_URL ?? 'https://data-seed-prebsc-1-s1.binance.org:8545'),
      });
      const signed = await client.signTransaction(payload as any);
      return signed.replace(/^0x/, '');
    },
  };
}

async function main() {
  const privateKey = process.env.BSC_CLIENT_PRIVATE_KEY as Hex;
  const serverUrl = process.env.SERVER_URL ?? 'http://127.0.0.1:8012';
  const endpoint = process.env.ENDPOINT ?? '/protected-bsc-testnet-coinbase';

  const wallet = createLocalWallet(privateKey);
  const signer = new EvmClientSigner(wallet);
  signer.setAddress(await wallet.getAddress());

  const x402 = new X402Client();
  x402.register('eip155:97', new ExactPermitEvmClientMechanism(signer));
  x402.register('eip155:97', new ExactEvmClientMechanism(signer));
  x402.registerPolicy(SufficientBalancePolicy);

  const client = new X402FetchClient(x402);
  const response = await client.get(`${serverUrl}${endpoint}`);
  console.log(response.status, response.statusText);

  const paymentHeader = response.headers.get('payment-response');
  if (paymentHeader) {
    const settle = decodePaymentPayload<SettleResponse>(paymentHeader);
    console.log(settle);
  }

  console.log(await response.text());
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});

File: legacy/examples/bsc-testnet-smoke/bsc_exact_client.py
#!/usr/bin/env python3
"""
Minimal BSC testnet exact client example using a local private key.
"""

import asyncio
import os

import httpx
from eth_account import Account
from eth_account.messages import encode_defunct, encode_typed_data

from bankofai.x402.clients import SufficientBalancePolicy, X402Client, X402HttpClient
from bankofai.x402.config import NetworkConfig
from bankofai.x402.mechanisms.evm.exact import ExactEvmClientMechanism
from bankofai.x402.mechanisms.evm.exact_permit import ExactPermitEvmClientMechanism
from bankofai.x402.signers.client import EvmClientSigner


class LocalEvmWallet:
    def __init__(self, private_key: str) -> None:
        key = private_key if private_key.startswith("0x") else f"0x{private_key}"
        self._account = Account.from_key(key)

    async def get_address(self) -> str:
        return self._account.address

    async def sign_message(self, message: bytes) -> str:
        signed = self._account.sign_message(encode_defunct(primitive=message))
        return "0x" + signed.signature.hex()

    async def sign_typed_data(self, full_data: dict) -> str:
        signed = Account.sign_message(
            encode_typed_data(full_message=full_data),
            private_key=self._account.key,
        )
        return "0x" + signed.signature.hex()

    async def sign_transaction(self, tx: dict) -> str:
        signed = self._account.sign_transaction(tx)
        raw_tx = signed.raw_transaction.hex()
        return raw_tx[2:] if raw_tx.startswith("0x") else raw_tx


async def main() -> None:
    private_key = os.environ["BSC_CLIENT_PRIVATE_KEY"]
    server_url = os.getenv("SERVER_URL", "http://127.0.0.1:8012")
    endpoint = os.getenv("ENDPOINT", "/protected-bsc-testnet-coinbase")

    signer_wallet = LocalEvmWallet(private_key)
    signer = EvmClientSigner(signer_wallet)
    signer.set_address(await signer_wallet.get_address())

    x402 = X402Client()
    x402.register(NetworkConfig.BSC_TESTNET, ExactPermitEvmClientMechanism(signer))
    x402.register(NetworkConfig.BSC_TESTNET, ExactEvmClientMechanism(signer))
    x402.register_policy(SufficientBalancePolicy)

    async with httpx.AsyncClient(timeout=120) as http_client:
        client = X402HttpClient(http_client, x402)
        response = await client.get(f"{server_url}{endpoint}")
        print(response.status_code)
        print(response.text[:500])


if __name__ == "__main__":
    asyncio.run(main())
